src/parser/core/modules/EventEmitter.ts

```diff
@@ -82,59 +89,73 @@ class EventEmitter extends Module {
     }
     return listener;
   }
-  _prependByCheck(listener, by) {
+
+  createByCheck<T extends Event>(by: number): ((event: Event) => boolean) | null {
     const requiresSelectedPlayer = by & SELECTED_PLAYER;
     const requiresSelectedPlayerPet = by & SELECTED_PLAYER_PET;
 
-    let check;
     if (requiresSelectedPlayer && requiresSelectedPlayerPet) {
-      check = event => this.owner.byPlayer(event) || this.owner.byPlayerPet(event);
+      return event => this.owner.byPlayer(event) || this.owner.byPlayerPet(event);
     } else if (requiresSelectedPlayer) {
-      check = event => this.owner.byPlayer(event);
+      return event => this.owner.byPlayer(event);
     } else if (requiresSelectedPlayerPet) {
-      check = event => this.owner.byPlayerPet(event);
+      return event => this.owner.byPlayerPet(event);
     } else {
+      return null;
+    }
+  }
+  _prependByCheck<T extends Event>(listener: EventListener<T>, by: number): EventListener<T> {
+    const check = this.createByCheck(by);
+    if (!check) {
       return listener;
     }
 
-    return function (event) {
+    return function(event) {
       if (check(event)) {
         listener(event);
       }
     };
   }
-  _prependToCheck(listener, to) {
+
+  createToCheck(to: number): ((event: Event) => boolean) | null {
     const requiresSelectedPlayer = to & SELECTED_PLAYER;
     const requiresSelectedPlayerPet = to & SELECTED_PLAYER_PET;
-
-    let check;
     if (requiresSelectedPlayer && requiresSelectedPlayerPet) {
-      check = event => this.owner.toPlayer(event) || this.owner.toPlayerPet(event);
+      return event => this.owner.toPlayer(event) || this.owner.toPlayerPet(event);
     } else if (requiresSelectedPlayer) {
-      check = event => this.owner.toPlayer(event);
+      return event => this.owner.toPlayer(event);
     } else if (requiresSelectedPlayerPet) {
-      check = event => this.owner.toPlayerPet(event);
+      return event => this.owner.toPlayerPet(event);
     } else {
+      return null;
+    }
+  }
+  _prependToCheck<T extends Event>(listener: EventListener<T>, to: number): EventListener<T> {
+    const check = this.createToCheck(to);
+    if (!check) {
       return listener;
     }
-
-    return function (event) {
+    return function(event: T) {
       if (check(event)) {
         listener(event);
       }
     };
   }
-  _prependSpellCheck(listener, spell) {
+
+  createSpellCheck(spell: SpellFilter): ((event: Event) => boolean) {
     if (spell instanceof Array) {
-      return function (event) {
-        if (event.ability && spell.some(s => s.id === event.ability.guid)) {
-          listener(event);
-        }
-      };
+      // @ts-ignore adding typechecking to <generic> event properties will be a lot of work
```

Comment:
Isn’t the Event type sufficient here?

Reply:
No, because `Event` does not define` .ability` (or .to, .by, etc); I had started normalizing the Event types by creating things like `AbilityEvent` `TargetedEvent`, etc, but things like event listeners are currently utilizing some of the lazy typing (by checking for undefined) and not being strict about it. So, we can either add things like `.ability .to .by` to Event and be kinda lazy, or we can get very strict all the way up to typechecking the EventListener you provide to `.addEventListener` to make sure it accepts the correct Event type parameter for the EventFilter you've given it.

Or, maybe there's an easier solution I'm missing!

Reply:
This is being resolved in a future PR; converting CLP and a lot of other core modules